src/Analyzers/CSharp/CodeFixes/NamespaceSync/CSharpNamespaceSyncCodeFixProvider.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.NamespaceSync
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.SyncNamespace), Shared]
+    internal class CSharpNamespaceSyncCodeFixProvider : CodeFixProvider
+    {
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpNamespaceSyncCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(cancellationToken => FixAsync(context.Document, cancellationToken)),
+                context.Diagnostics);
+
+            return Task.CompletedTask;
+        }
+
+        protected static async Task<Solution> FixAsync(Document document, CancellationToken cancellationToken)
+        {
+            // Use the Renamer.RenameDocumentAsync API to sync namespaces in the document. This allows
+            // us to keep in line with the sync methodology that we have as a public API and not have 
+            // to rewrite or move the complex logic. RenameDocumentAsync is designed to behave the same
+            // as the intent of this analyzer/codefix pair.
+            var currentFolders = document.Folders;
+            var documentWithNoFolders = document.WithFolders(new string[0]);
+            var renameActionSet = await Renamer.RenameDocumentAsync(
+                documentWithNoFolders,
+                documentWithNoFolders.Name,
+                newDocumentFolders: currentFolders,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return await renameActionSet.UpdateSolutionAsync(documentWithNoFolders.Project.Solution, cancellationToken).ConfigureAwait(false);
+        }
+
+        public override FixAllProvider? GetFixAllProvider()
```

Comment:
this all feels unnecessary.  why do we need a custom fix all provider here?  why would the batch provider not be sufficient for fix all?  since there is max one fix per file, we can just use the default behavior to fix each file independently and merge hte result.